File: port/src/cbits/Cocoa/Timer.m
#include "Timer.h"
#include "Handlers_stub.h"
#include "Internals.h"

TimerHandle osCreateTimer(int msecs)
{
	NSTimer *timer = 
		[[NSTimer alloc] initWithFireDate: NULL
		                         interval: msecs*1000
		                           target: [NSApp delegate]
		                         selector: @selector(timerClicked:)
		                         userInfo: NULL
		                          repeats: YES];
	return timer;
}

void osDestroyTimer(TimerHandle timer)
{
	printf("osDestroyTimer\n");
}

void osSetTimerInterval(TimerHandle timer, int msecs)
{
	printf("osSetTimerInterval\n");
};

int osGetTimerInterval(TimerHandle timer)
{
	printf("osGetTimerInterval\n");
	return 0;
};

void osSetTimerEnabled(TimerHandle timer, BOOL enabled)
{
	printf("osSetTimerEnabled\n");
};

BOOL osGetTimerEnabled(TimerHandle timer)
{
	printf("osGetTimerEnabled\n");
	return 0;
};

File: port/src/cbits/GTK/Timer.c
#include "Timer.h"
#include "Handlers_stub.h"
#include "Internals.h"

static gboolean osTimerProc(gpointer data)
{
  handleTimer((TimerHandle) data);
  return TRUE;
}

TimerHandle osCreateTimer(int msecs)
{
	TimerHandle timer = rmalloc(sizeof(*timer));
	timer->interval = msecs;
	timer->enabled  = TRUE;
	timer->id = (msecs > 0) ? gtk_timeout_add(msecs, osTimerProc, timer) : 0;
	return timer;
}

void osDestroyTimer(TimerHandle timer)
{
	if (timer!=NULL)
	{
	    handleTimerDestroy(timer);
		if (timer->id > 0)
			gtk_timeout_remove(timer->id);
		rfree(timer);
	}
}

void osSetTimerInterval(TimerHandle timer, int msecs)
{
	if (timer->interval != msecs)
	{
		timer->interval = msecs;
		if (timer->enabled)
		{
			if (timer->id > 0)
				gtk_timeout_remove(timer->id);
			timer->id = (msecs > 0) ? gtk_timeout_add(msecs, osTimerProc, timer) : 0;
		}
	}
};

int osGetTimerInterval(TimerHandle timer)
{
	return timer->interval;
};

void osSetTimerEnabled(TimerHandle timer, BOOL enabled)
{
	timer->enabled = enabled;
	if (timer->enabled)
	{
		if (timer->id <= 0 && timer->interval > 0)
			timer->id = gtk_timeout_add(timer->interval, osTimerProc, timer);
	}
	else
	{
		if (timer->id > 0)
		  gtk_timeout_remove(timer->id);
		timer->id = 0;
	}
};

BOOL osGetTimerEnabled(TimerHandle timer)
{
	return timer->enabled;
};
